python/ray/data/_internal/arrow_ops/transform_pyarrow.py

```diff
@@ -74,30 +75,46 @@ def unify_schemas(
             if pa.types.is_list(col_type) and pa.types.is_null(col_type.value_type):
                 cols_with_null_list.add(col_name)
 
-    if any(isinstance(type_, pyarrow.ExtensionType) for type_ in schemas[0].types):
-        # If we have pyarrow extension types that may potentially be variable shaped,
-        # examine the first schema to gather the columns that need type conversions.
-        for col_field in schemas[0]:
-            col_name, col_type = col_field.name, col_field.type
-            tensor_array_types = [
-                s.field(col_name).type
-                for s in schemas
-                if isinstance(s.field(col_name).type, pyarrow.ExtensionType)
-            ]
-            if ArrowTensorType._need_variable_shaped_tensor_array(tensor_array_types):
-                if isinstance(tensor_array_types[0], ArrowVariableShapedTensorType):
-                    new_type = tensor_array_types[0]
-                elif isinstance(tensor_array_types[0], ArrowTensorType):
-                    new_type = ArrowVariableShapedTensorType(
-                        dtype=tensor_array_types[0].scalar_type,
-                        ndim=len(tensor_array_types[0].shape),
-                    )
-                else:
-                    raise ValueError(
-                        "Detected need for variable shaped tensor representation, "
-                        f"but schema is not ArrayTensorType: {tensor_array_types[0]}"
-                    )
-                schema_field_overrides[col_name] = new_type
+    arrow_tensor_types = (ArrowVariableShapedTensorType, ArrowTensorType)
```

Comment:
can you explain what this change is doing?

Reply:
Basically this is a refactor of the unify_schemas function, which serves a similar purpose to the PyArrow unify_schemas. Its job is to take the schemas of the tables to be concatenated and produce a "unified" schema which contains the correct types for all of the columns of the constituent tables.

The previous code checked only the first table in the tables to concatenate, so it would behave weirdly if the first table weren't a tensor type and there was one further along by creating a schema where the tensor types were just erased (the error would be caught elsewhere, though).

However, for the object array, we need to support the case where the first table has a column of type, say, int64, and the second table has a column of type SomeObject. The old code would simply set the type to int64, which would cause an error. The refactored code will ensure that the int64s get pickled (yes, this is not very efficient, but this is the only way to support  So this new code scans all of the table schemas, finds which columns have any table that contain tensors or objects, and handles them accordingly.

This is basically type promotion for the ArrowTensorArray and the ArrowVariableShapedTensorArray and likewise for the ArrowPythonObjectArray and other types. Should we set the promotion options differently too?

Reply:
Thanks for the detailed explanation! Can you also update the docstring of this method to reflect this change?
also do you mind explaining again what you mean by "set the promotion options differently"?